### ISSUE TITLE
Discourage use of certain npm packages

### ISSUE THREAD

**Issue by jimmywarting** [xojs/xo#550]
- **inherits** node docs recommends to use class extends instead
- **mkdirp** node v10.12 mkdir supports recursive option
- **xtend** use `Object.assign` or spread `{ ...obj }`
- **rimraf** node supports recursive option now
- **pad-left**, **pad-right**, **left-pad**, **right-pad**, **pad** just use `"".padStart()` and `"".padEnd()`
- **safe-buffer** not needed anymore
- **array-flatten** use flat instead or some other polyfill
- **request** been deprecated

**Comment by sindresorhus** [comment]
I agree with all of these. Let's leave it open for a while to gather more ideas.

**Comment by sindresorhus** [comment]
- `object-assign` - Same reason as `xtend`.

**Comment by jimmywarting** [comment]
Added object-assign to the list. 
I just wish to see fewer dependencies and smaller builds/footprint. JS is what makes the web slow - not larger images
Unfortally not all packages gets deprecated - only discontinued

**Comment by papb** [comment]
> **rimraf** node supports recursive option now

**Comment by jimmywarting** [comment]
Are you really sure Node.js' new option entirely replaces rimraf, including all the crazy hacks it has to make it work on Windows?

I do not know about all the hacks but at least this exist in NodeJs and if there is some window bugs out there then i would have expect some of them to be fixed in v16

```js
// Added in: v14.14.0
fs.rm(path {recursive: true}, callback) 
fsPromises.rm(path {recursive: true})
```

**Comment by sindresorhus** [comment]
Eventually, I'd also like to discourage `stream` in favor of `stream/web`: https://github.com/nodejs/node/blob/master/doc/api/webstreams.md

**Comment by jimmywarting** [comment]
Agree with you, Stream is going to be a tuff one to get ppl to stop using.
But i agree that it would be best to have something that works in Deno, Node & browser the same way and follows a spec'ed standard

for now i just use the async iterator part of node:streams

**Comment by jimmywarting** [comment]
Saw this in a yarn log
```js
warning @angular-devkit/build-angular > webpack-dev-server > url > querystring@0.2.0: The querystring is deprecated
```

Should we do something with `url`?

**Comment by sindresorhus** [comment]
> Should we do something with url?

There's no way for us to know whether it's the built-in `url` or the `url` package. I also don't think a lot of projects use the `url` package.

**Comment by Richienb** [comment]
A list of discouraged packages should be compiled in a seperate repository than the main xo one so it can move faster and so there is a single centralised place to add explainations and migration guides.

**Comment by sindresorhus** [comment]
@Richienb That will just make it more difficult to maintain. Every added package is a breaking change as it causes a new lint error, so it would need an XO version bump anyway. We can consider doing that in the future if the list grows huge.

**Comment by jimmywarting** [comment]
> @Richienb That will just make it more difficult to maintain. Every added package is a breaking change as it causes a new lint error, so it would need an XO version bump anyway. We can consider doing that in the future if the list grows huge.

hmm, i think the list could change quite a lot. Would be annoying to have to bump major every time it change.
maybe it should just be a warning instead of an error?

Also what about checking sub dependencies also? doing so could encourage developers to send issues/pr upstream in the hope to reduce the number of warning messages

**Comment by sindresorhus** [comment]
> Would be annoying to have to bump major every time it change.
maybe it should just be a warning instead of an error?

Warnings don't really work. They are too easily ignored. And the user could have changed it to an error themselves, so we would still need a major bump (minor in XO's case because we're pre-1.0.0).

> Also what about checking sub dependencies also? doing so could encourage developers to send issues/pr upstream in the hope to reduce the number of warning messages

That could potentially be the next step, but I think we should try out direct dependencies first.
